Sources/_MatchingEngine/Regex/Parse/SourceLocation.swift

```diff
@@ -0,0 +1,135 @@
+extension Source {
+  /// The location in the input of a parsed entity, presented as a region over the input
+  public struct Location: Hashable {
+    public var range: Range<Source.Position>
+
+    public var start: Source.Position { range.lowerBound }
+
+    /// The open end
+    public var end: Source.Position { range.upperBound }
+
+    public init(_ r: Range<Source.Position>) {
+      self.range = r
+    }
+    public init<R: RangeExpression>(
+      _ r: R, in input: Source
+    ) where R.Bound == Source.Position {
+      self.init(r.relative(to: input.input))
+    }
+
+    /// A "fake" location, sometimes useful for tooling purposes
+    public static var fake: Self {
+      .init("".startIndex ..< "".startIndex)
+    }
+    public var isFake: Bool { self == Self.fake }
+    public var isReal: Bool { !isFake }
+  }
+}
+public typealias SourceLocation = Source.Location
+
+extension Source {
+  var currentPosition: Position { bounds.lowerBound }
+}
+
+extension Source {
+  /// An error with source location info
+  public struct Err<E: Error>: Error {
+    public let error: E
+    public let location: SourceLocation
+
+    init(_ e: E, _ r: SourceLocation) {
+      self.error = e
+      self.location = r
+    }
+    public init(_ v: E, _ r: Range<Source.Position>) {
+      self.error = v
+      self.location = Location(r)
+    }
+
+  }
+
+  /// Located value: a value wrapped with a source range
+  ///
+  /// Note: source location is part of value identity, so that the same
+  /// e.g. `Character` appearing twice can be stored in a data structure
+  /// distinctly. To ignore source locations, use `.value` directly.
+  public struct Loc<T: Hashable>: Hashable {
```

Comment:
The existence of both `Source.Location` and `Source.Loc` makes it less readable as the latter could be misinterpreted as the former (esp. for people coming from the compiler repo, i.e. `SourceLoc`).  How about using the full name `Located<T>`

Reply:
I agree for public API (even public for the parser).